### PR TITLE
All labels benchmark

### PR DESCRIPTION
**Pull request by ben-heil**
This PR contains the new design for the benchmark as laid out [here](https://github.com/greenelab/saged/issues/3#issuecomment-666509142). Namely, lots of config files for controlling how things run, and general benchmark scripts that are agnostic about what type of model/dataset they are running. Unfortunately this led to a lot of code changes. I'm sorry for the size of the PR.

Changes:
- Created lots of config files
- Wrote the first benchmark, `all_label_comparison`, that trains a model to do a multiclass prediction on all labels in the dataset
- Made the infrastructure for datasets containing both labeled and unlabeled data
- Updated the `models.py` and `datasets.py` classes to take config files information
- Added more test cases to test the new functionality

### CODE REVIEW

File: saged/all_label_comparison.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                        help='The yaml formatted model configuration file for an unsupervised '
+                             'model. If omitted, the benchmark will use the original features '
+                             'for training. For more information about this file read the '
+                             'comments in the example_model.yml file',
+                        default=None)
+    parser.add_argument('--neptune_config',
+                        help='A yaml formatted file containing init information for '
+                             'neptune logging')
+    parser.add_argument('--seed',
+                        help='The random seed to be used in the experiment',
```
Comment (jjc2718) [inline saged/all_label_comparison.py]: Might be good to note that this seed only controls the way the data is split, not the model initialization etc. (looks like the seed for the Pytorch model is set separately in the relevant model config).

File: saged/all_label_comparison.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    print('Loading all data')
+    all_data = MixedDatasetClass.from_config(**dataset_config)
+    print('Subsetting labeled data')
+    labeled_data = all_data.get_labeled()
+    print('Subsetting unlabeled data')
+    unlabeled_data = all_data.get_unlabeled()
+    print('Splitting data')
+
+    # Get fivefold cross-validation splits
+    labeled_splits = labeled_data.get_cv_splits(num_splits=5, seed=args.seed)
```
Comment (jjc2718) [inline saged/all_label_comparison.py]: You could make the number of splits a command-line parameter as well, or define it in a config file somewhere. I doubt you'll need to change it much, if at all, but it probably wouldn't hurt to make it more accessible.

File: saged/datasets.py [hunk 3]
```diff
@@ -320,34 +366,20 @@ def __len__(self) -> int:
...
-        -------
-        sample: The gene expression data for the given index in a genes x 1 array
-        """
-        sample_id = self.get_samples()[idx]
-        sample = self.current_expression[sample_id].values
-        return sample
-
-    def get_all_data(self) -> Tuple[np.array, np.array]:
+        new_data: The variable x samples array to overwrite the old data with. The first dimension
+                  can be anything, but the sample dimension must remain the same as in the old data
```
Comment (jjc2718) [inline saged/datasets.py]: It's not just that the dimension has to be the same, but the columns have to be in the same order too, right?

File: saged/datasets.py [hunk 4]
```diff
@@ -731,16 +943,59 @@ def from_paths(class_object,
...
+                # https://stackoverflow.com/questions/19125091/pandas-merge-how-to-avoid-duplicating-columns
+                cols_to_use = current_dataframe.columns.difference(combined_dataframe.columns)
+
+                # Combine dataframes
+                combined_dataframe = pd.merge(combined_dataframe,
+                                              current_dataframe[cols_to_use],
+                                              left_index=True,
+                                              right_index=True,
+                                              how='outer')
+                # Combine labels (double start expands a dictionary into key-value pairs)
```
Comment (jjc2718) [inline saged/datasets.py]:
```suggestion
                # Combine labels (double star expands a dictionary into key-value pairs)
```
Also, do you want this comment in the function in `RefineBioUnlabeledDataset` too?

File: saged/datasets.py [hunk 5]
```diff
@@ -952,3 +1223,266 @@ def train_test_split(self,
...
```
Comment (jjc2718) [inline saged/datasets.py]:
```suggestion
                # Combine labels (double star expands a dictionary into key-value pairs)
```

File: saged/models.py [hunk 6]
```diff
@@ -185,11 +201,16 @@ class LogisticRegression(ExpressionModel):
...
+                 seed: int,
+                 **kwargs,
+                 ) -> None:
         """
         The initializer the LogisticRegression class
 
-        seed: The random seed to be used by the model
+        Arguments
+        ---------
+        seed: The random seed ot use in training
```
Comment (jjc2718) [inline saged/models.py]:
```suggestion
        seed: The random seed to use in training
```

File: saged/models.py [hunk 7]
```diff
@@ -185,11 +201,16 @@ class LogisticRegression(ExpressionModel):
     base acceptable functions for models in this module's benchmarking code
     """
 
-    def __init__(self, seed: int = 42) -> None:
+    def __init__(self,
+                 seed: int,
+                 **kwargs,
+                 ) -> None:
         """
         The initializer the LogisticRegression class
```
Comment (jjc2718) [inline saged/models.py]:
```suggestion
        The initializer of the LogisticRegression class
```

File: saged/models.py [hunk 8]
```diff
@@ -410,8 +501,6 @@ def fit(self, dataset: LabeledDataset) -> "PytorchSupervised":
             self.loss_fn = self.loss_class(weight=None)
 
         if log_progress:
-            # Set up the neptune experiment
-            utils.initialize_neptune(self.config)
```
Comment (jjc2718) [inline saged/models.py:414]: Is this function still being used anywhere?

File: test/test_datasets.py [hunk 9]
```diff
@@ -344,3 +386,95 @@ def test_subset_samples_to_labels(labeled_datasets, labels):
...
+
+def test_get_samples(all_datasets):
+    # Get sample names
+    current_dir = os.path.dirname(os.path.abspath(__file__))
+    expression_path = os.path.join(current_dir, 'data/test_expression.tsv')
+    with open(expression_path) as expression_file:
+        true_samples = set(expression_file.readline().strip().split('\t'))
+
+    for dataset in all_datasets:
+        print(type(dataset))
```
Comment (jjc2718) [inline test/test_datasets.py]: Do you want to leave this print statement in?

**Review by jjc2718 — APPROVED** [review]
Looks good! My comments are mainly just typo fixes and small questions.

I assume you haven't actually run the benchmark script yet (so this is just a proof of concept/prototype of what you'll be running)? Or have you?

File: saged/all_label_comparison.py (hunk 1 above)
Comment (ben-heil) [reply saged/all_label_comparison.py]: Good point

File: saged/all_label_comparison.py (hunk 2 above)
Comment (ben-heil) [reply saged/all_label_comparison.py]: Will do

File: saged/datasets.py (hunk 3 above)
Comment (ben-heil) [reply saged/datasets.py]: That's true, updated the comment

File: saged/datasets.py (hunk 4 above)
Comment (ben-heil) [reply saged/datasets.py]: Went ahead and added it to `RefineBioUnlabeledDataset`

File: saged/models.py (hunk 6 above)
Comment (ben-heil) [reply saged/models.py]: Eek, sorry for all the typos

File: saged/models.py (hunk 8 above)
Comment (ben-heil) [reply saged/models.py:414]: Yeah, I moved it to `all_labels_comparison`, with the idea that each benchmark will be responsible for setting up its own connection with neptune. It also makes more it so that it's unnecessary to pass all the neptune configuration information to every model

File: test/test_datasets.py (hunk 9 above)
Comment (ben-heil) [reply test/test_datasets.py]: Good catch, went ahead and removed the debug print statements
